### PR TITLE
docment resize crop pad

### PR DESCRIPTION
**Pull request by dienhoa**
Docment  
- Resize with crop, pad or squish
- RandomCrop
- Resize
- RandomResizedCrop
- RatioResize

in vision.augment, issue: https://github.com/muellerzr/fastai-docment-sprint/issues/29

### CODE REVIEW

File: fastai/vision/augment.py [hunk 1]
```diff
@@ -259,14 +278,24 @@ def encodes(self, x:(Image.Image,TensorBBox,TensorPoint)):
 class RandomResizedCrop(RandTransform):
     "Picks a random scaled crop of an image and resize it to `size`"
     split_idx,order = None,1
-    def __init__(self, size, min_scale=0.08, ratio=(3/4, 4/3), resamples=(Image.BILINEAR, Image.NEAREST),
-                 val_xtra=0.14, max_scale=1., **kwargs):
+    def __init__(self,
+         size:(int, (int,int)), # Final size (after crop and resize), duplicated if one value is specified,,
```
Comment (muellerzr) [inline fastai/vision/augment.py:282]:
```suggestion
         size:(int, tuple), # Final size, duplicated if one value is specified,,
```

File: fastai/vision/augment.py [hunk 2]
```diff
@@ -259,14 +278,24 @@ def encodes(self, x:(Image.Image,TensorBBox,TensorPoint)):
...
+         min_scale:float=0.08, # Min scale of the crop
```
Comment (muellerzr) [inline fastai/vision/augment.py:283]:
```suggestion
         min_scale:float=0.08, # Minimum scale of the crop
```
We already have `min`. Make it more verbose by saying what it is

File: fastai/vision/augment.py [hunk 3]
```diff
@@ -259,14 +278,24 @@ def encodes(self, x:(Image.Image,TensorBBox,TensorPoint)):
...
+         ratio:(float, float)=(3/4, 4/3), # Range of width/height of the output
```
Comment (muellerzr) [inline fastai/vision/augment.py:284]:
```suggestion
         ratio:(float, float)=(3/4, 4/3), # Range of (width, height) of the output
```

**Comment by dienhoa** [comment]
I commit a new version of this PR with your suggestions @muellerzr. Thanks.

File: fastai/vision/augment.py (hunk 3 above)
Comment (dienhoa) [reply fastai/vision/augment.py]: I change it to # Range of width over height of the output. I think it's clearer this way right?

**Comment by marii-moe** [comment]
@dienhoa I noticed you added split_idx=1 to this code `ropped = crop(img, split_idx=1)`

This is stopping the example images from being augmented.

File: fastai/vision/augment.py [hunk 4]
```diff
@@ -259,14 +278,24 @@ def encodes(self, x:(Image.Image,TensorBBox,TensorPoint)):
...
     split_idx,order = None,1
-    def __init__(self, size, min_scale=0.08, ratio=(3/4, 4/3), resamples=(Image.BILINEAR, Image.NEAREST),
-                 val_xtra=0.14, max_scale=1., **kwargs):
+    def __init__(self,
+         size:(int, tuple), # Final size, duplicated if one value is specified,,
+         min_scale:float=0.08, # Minimum scale of the crop
+         ratio:(float, float)=(3/4, 4/3), # Range of width over height of the output
+         resamples=(Image.BILINEAR, Image.NEAREST), # Pillow `Image` resample mode, resamples[1] for mask
+         val_xtra:float=0.14, # Ratio of extra pixels in the validation set
+         max_scale:float=1., # Maximum scale of the crop
```
Comment (marii-moe) [inline fastai/vision/augment.py:287]:
```suggestion
         min_scale:float=0.08, # Minimum scale of the crop, in relation to image area
         ratio:(float, float)=(3/4, 4/3), # Range of width over height of the output
         resamples=(Image.BILINEAR, Image.NEAREST), # Pillow `Image` resample mode, resamples[1] for mask
         val_xtra:float=0.14, # For validation set, ratio of size extra pixels to use in a squish before cropping to size.
         max_scale:float=1., # Maximum scale of the crop, in relation to image area
```

**Review by marii-moe — CHANGES_REQUESTED** [review]
Requested some changes to get to the heart of some of the parameters. I am having a hard time wording val_xtra as well, so feel free to come up with something better :)

**Comment by dienhoa** [comment]
> @dienhoa I noticed you added split_idx=1 to this code `cropped = crop(img, split_idx=1)`
> 
> This is stopping the example images from being augmented.

Oh sorry, I did some experiments and forget to clean the notebook. Will change it back to the original

File: fastai/vision/augment.py (hunk 4 above)
Comment (dienhoa) [reply fastai/vision/augment.py:287]: For the val_xtra, I don't quite understand why we need to squish to a bigger dimension then crop down to size? It just means we cut the pixels at the border

Comment (dienhoa) [reply fastai/vision/augment.py:287]: I think the model is trained with crop image then it is better if the validation image is going through some sort of cropping too.

For the document, How about,  `the ratio of size at the edge cut out in the validation set`? 

Thanks

Comment (marii-moe) [reply fastai/vision/augment.py:287]: First comment: Because depending on the squish algorithm it needs to have some extra border so that is can get similar resizing artifacts to that which we see during training. We resize to be a bit bigger than our final crop so that most of the image will be included in the crop. We want a large center crop during validation so that the model can see most of the image at test time. 

Second comment: For validation we resize to approximately:             
```
xtra =largest_img_deminsion * self.val_xtra
self.final_size = (self.size[0]+xtra, self.size[1]+xtra)
```
First for validation we resize to self.final_size. We then crop to the regular `size` parameter. For training the transform randomly does the resize, and crop. For validation this is just an easier version that means our model will have an easier time at validation/test time, giving us better results in production in the end. 

`the ratio of size at the edge cropped out in the validation set`
switch cut for cropped and we have a winner.

**Comment by dienhoa** [comment]
Thanks for your detailed explanation @marii-moe. I commit a new version with suggestions

File: fastai/vision/augment.py [hunk 5]
```diff
@@ -191,12 +195,19 @@ def encodes(self, x:(Image.Image,TensorBBox,TensorPoint)):
...
         size = _process_sz(size)
         store_attr()
         super().__init__(**kwargs)
 
-    def before_call(self, b, split_idx):
+    def before_call(self,
+        b,
+        split_idx:int # Index of the train/valid dataset
+    ):
+        "Randomly positioning crop if train dataset (split_idx=0) else center crop"
```
Comment (muellerzr) [inline fastai/vision/augment.py:210]:
```suggestion
        "Randomly positioning crop if `split_idx==0` else center crop"
```

**Review by muellerzr — APPROVED** [review]
One small nit, @marii-moe can you also give it one last look otherwise should be good to go!

File: fastai/vision/augment.py (hunk 5 above)
Comment (mollybeavers) [reply fastai/vision/augment.py:210]: @muellerzr why not just "Randomly positioning crop if training, otherwise center crop "

Then we leave the `split_idx` information in the document for the param, as `split_idx` is more of an implementation detail of how Dataloaders work with transforms. I wouldn't expect most users to know that `split_idx==0` means we are working with the training set.

Comment (marii-moe) [reply fastai/vision/augment.py:210]: @muellerzr why not keep train dataset instead? I think that would be more familiar to the majority of users, making the documentation easier to digest. Knowing what split_idx is can be left to the documentation for that specific param.

**Review by marii-moe — COMMENTED** [review]
Just need @muellerzr response to the above conversation, otherwise LGTM.

File: fastai/vision/augment.py (hunk 5 above)
Comment (marii-moe) [reply fastai/vision/augment.py:210]: @dienhoa had a conversation with zach, go with my recommendation and we can get this merged. So use: "Randomly positioning crop if train dataset else center crop"

Comment (dienhoa) [reply fastai/vision/augment.py]: Done. Thanks @marii-moe !

**Comment by dienhoa** [comment]
github question. If now I rebase my branch with @muellerzr  master to resolve conflict, does it pollute everything we've done here?

**Comment by dienhoa** [comment]
I solved the conflicts and have some notes:

- I failed to use `git rebase` and found that `git merge` is much easier to use in case of conflicts. I'm not sure what happened but after using rebase and fixing all conflicts, I run `git rebase --continue` and my Jupiter notebook crashed. Didn't have time to investigate more on it.
- There are a lot of conflicts just because of the Image output change. It's inevitable because we have some sort of random process. I think we can not do anything in nbdev to solve it right?

**Comment by dienhoa** [comment]
> @dienhoa There is this to try: https://nbdev.fast.ai/merge.html#nbdev_fix_merge
> 
> Otherwise I can take a look at it tomorrow.

I did run `nbdev_fix_merge` already. The conflict is solved in my last commit (you can see that all CI jobs passed). Just my thinking that we will always have conflicts because the output image always change

**Comment by marii-moe** [comment]
@dienhoa sorry I misunderstood. I will go ahead and merge then.

**Review by marii-moe — APPROVED** [review]
LGTM
